core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map.Entry;
+
+import org.reactivestreams.Publisher;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+
+@SuppressWarnings("checkstyle:OverloadMethodsDeclarationOrder")
+abstract class AbstractHttpMessageBuilder {
+
+    @Nullable
+    private HttpData content;
+
+    @Nullable
+    private Publisher<? extends HttpData> publisher;
+
+    @Nullable
+    private HttpHeadersBuilder httpTrailers;
+
+    @Nullable
+    final HttpData content() {
+        return content;
+    }
+
+    @Nullable
+    final Publisher<? extends HttpData> publisher() {
+        return publisher;
+    }
+
+    @Nullable
+    final HttpHeadersBuilder httpTrailers() {
+        return httpTrailers;
+    }
+
+    abstract HttpHeadersBuilder headersBuilder();
+
+    AbstractHttpMessageBuilder header(CharSequence name, Object value) {
+        headersBuilder().addObject(requireNonNull(name, "name"),
+                                   requireNonNull(value, "value"));
+        return this;
+    }
+
+    AbstractHttpMessageBuilder headers(
+            Iterable<? extends Entry<? extends CharSequence, String>> headers) {
+        requireNonNull(headers, "headers");
+        headersBuilder().add(headers);
+        return this;
+    }
+
+    AbstractHttpMessageBuilder content(String content) {
```

Comment:
I didn't request it because Javadoc of subclasses has request or response in the sentence depending on its type. But either is fine to me. 😉

Reply:
> I didn't request it because Javadoc of subclasses has request or response in the sentence depending on its type.

The reason why I wrote Javadoc in subclasses is this.

Reply:
OK to leave it as it is.